File: rust/lib/src/msgs/account.rs
use serde::Deserialize;
use crate::common::order_status::OrderStatus;
use crate::common::side::Side;

// ─────────────────────────────────────────────────────────────────────────────
// Margin
// ─────────────────────────────────────────────────────────────────────────────

/// Account margin information (from WebSocket `marginUpdate` / `accountSnapshot`).
#[derive(Debug, Clone, Default, Deserialize)]
#[allow(unused)]
pub struct Margin {
    #[serde(rename = "totalBalance")]
    pub total_balance: f64,
    #[serde(rename = "availableBalance")]
    pub available_balance: f64,
    #[serde(rename = "marginUsed")]
    pub margin_used: f64,
    pub notional: f64,
    #[serde(rename = "realizedPnl")]
    pub realized_pnl: f64,
    #[serde(rename = "unrealizedPnl")]
    pub unrealized_pnl: f64,
    pub fees: f64,
    pub funding: f64,
}

// ─────────────────────────────────────────────────────────────────────────────
// Positions
// ─────────────────────────────────────────────────────────────────────────────

/// Position information.
///
/// Deserializes from both WebSocket and HTTP payloads:
/// - WS uses `"symbol"`, HTTP uses `"coin"` → `#[serde(alias)]`
/// - Fields only present on WS default to `0.0` when absent (HTTP).
#[derive(Debug, Clone, Deserialize)]
#[allow(unused)]
pub struct PositionInfo {
    #[serde(alias = "coin")]
    pub symbol: String,
    pub size: f64,
    pub price: f64,
    #[serde(rename = "fairPrice", default)]
    pub fair_price: f64,
    #[serde(default)]
    pub notional: f64,
    #[serde(rename = "realizedPnl", default)]
    pub realized_pnl: f64,
    #[serde(rename = "unrealizedPnl", default)]
    pub unrealized_pnl: f64,
    #[serde(default)]
    pub leverage: f64,
    #[serde(rename = "liquidationPrice", default)]
    pub liquidation_price: f64,
    #[serde(rename = "maintenanceMargin", default)]
    pub maintenance_margin: f64,
}

// ─────────────────────────────────────────────────────────────────────────────
// Order State
// ─────────────────────────────────────────────────────────────────────────────

/// Resting or historical order state.
///
/// Deserializes from both WebSocket and HTTP payloads:
/// - `vwap`, `reduce_only`, `tif` are HTTP-only (default when absent).
/// - `error` / `reason` is WS-only (default when absent).
#[derive(Debug, Clone, Deserialize)]
#[allow(unused)]
pub struct OrderState {
    #[serde(default)]
    pub timestamp: u64,
    pub symbol: String,
    #[serde(rename = "orderId")]
    pub order_id: String,
    pub status: OrderStatus,
    #[serde(rename = "isBuy")]
    pub side: Side,
    pub price: f64,
    pub size: f64,
    #[serde(rename = "filledSize", default)]
    pub filled_size: f64,
    #[serde(rename = "originalSize", default)]
    pub original_size: f64,
    #[serde(rename = "maker", default)]
    pub is_maker: bool,
    #[serde(rename = "reason", default)]
    pub error: Option<String>,
    /// Volume-weighted average fill price (HTTP only).
    #[serde(default)]
    pub vwap: f64,
    /// Whether this is a reduce-only order (HTTP only).
    #[serde(rename = "reduceOnly", default)]
    pub reduce_only: bool,
    /// Time-in-force as a string, e.g. "gtc" (HTTP only).
    #[serde(default)]
    pub tif: Option<String>,
}

// ─────────────────────────────────────────────────────────────────────────────
// Fills
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Deserialize)]
#[allow(unused)]
pub struct Fill {
    pub timestamp: u64,
    #[serde(alias = "coin")]
    pub symbol: String,
    #[serde(rename = "orderId")]
    pub order_id: String,
    pub price: f64,
    pub size: f64,
    #[serde(rename = "isBuy")]
    pub side: Side,
    #[serde(rename = "maker", default)]
    pub is_maker: bool,
}

// ─────────────────────────────────────────────────────────────────────────────
// Leverage Setting
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Deserialize)]
#[allow(unused)]
pub struct LeverageSetting {
    #[serde(alias = "coin")]
    pub symbol: String,
    pub leverage: f64,
}

// ─────────────────────────────────────────────────────────────────────────────
// Full Account (HTTP response)
// ─────────────────────────────────────────────────────────────────────────────

/// The inner payload of a `fullAccount` HTTP response.
///
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[allow(unused)]
pub struct AccountData {
    pub positions: Vec<PositionInfo>,
    pub open_orders: Vec<OrderState>,
    pub margin: Margin,
    pub leverage_settings: Vec<LeverageSetting>,
}


//
// Unit tests
//

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_order_state_rejected_risk_limit() {
        let json = r#"{
            "status": "rejectedRiskLimit",
            "symbol": "BTC-USD",
            "orderId": "EF2bxQ5pp3CDFAwRi44ExXb32sRmByByYxjwLYBfvRKQ",
            "price": 100001.37,
            "originalSize": 0.02474,
            "size": 0.02474,
            "filledSize": 0.0,
            "vwap": 0.0,
            "isBuy": false,
            "maker": true,
            "tif": "gtc",
            "timestamp": 1770918312787284000,
            "reason": "no oracle / fair price reference yet for: BTC-USD"
        }"#;

        let order: OrderState = serde_json::from_str(json).unwrap();

        assert_eq!(order.symbol, "BTC-USD");
        assert_eq!(order.order_id, "EF2bxQ5pp3CDFAwRi44ExXb32sRmByByYxjwLYBfvRKQ");
        assert_eq!(order.status, OrderStatus::RejectedRiskLimit);
        assert_eq!(order.side, Side::Sell);
        assert!((order.price - 100001.37).abs() < 1e-6);
        assert!((order.original_size - 0.02474).abs() < 1e-8);
        assert!((order.size - 0.02474).abs() < 1e-8);
        assert_eq!(order.filled_size, 0.0);
        assert!(order.is_maker);
        assert_eq!(order.timestamp, 1770918312787284000);
        assert_eq!(
            order.error.as_deref(),
            Some("no oracle / fair price reference yet for: BTC-USD")
        );

        // status helpers
        assert!(order.status.is_terminal());
        assert!(order.status.is_rejected());
    }

    #[test]
    fn test_order_state_from_ws_envelope() {
        let json = r#"{
            "type": "account",
            "data": {
                "type": "orderUpdate",
                "status": "rejectedRiskLimit",
                "symbol": "BTC-USD",
                "orderId": "EF2bxQ5pp3CDFAwRi44ExXb32sRmByByYxjwLYBfvRKQ",
                "price": 100001.37,
                "originalSize": 0.02474,
                "size": 0.02474,
                "filledSize": 0.0,
                "vwap": 0.0,
                "isBuy": false,
                "maker": true,
                "tif": "gtc",
                "timestamp": 1770918312787284000,
                "reason": "no oracle / fair price reference yet for: BTC-USD"
            },
            "topic": "account.2bZfxVQtWdd8qAWJ4Xyq43cnej9zqMNyuh7HHxTNan8j"
        }"#;

        // Parse the same way the actor does: from data["data"]
        let envelope: serde_json::Value = serde_json::from_str(json).unwrap();
        let order: OrderState =
            serde_json::from_value(envelope["data"].clone()).unwrap();

        assert_eq!(order.symbol, "BTC-USD");
        assert_eq!(order.status, OrderStatus::RejectedRiskLimit);
        assert_eq!(order.side, Side::Sell);
        assert!(order.error.is_some());
    }

    #[test]
    fn test_order_state_resting_no_reason() {
        let json = r#"{
            "status": "resting",
            "symbol": "ETH-USD",
            "orderId": "abc123",
            "price": 3200.0,
            "originalSize": 1.0,
            "size": 1.0,
            "filledSize": 0.0,
            "isBuy": true,
            "maker": true,
            "timestamp": 1770918312787284000
        }"#;

        let order: OrderState = serde_json::from_str(json).unwrap();

        assert_eq!(order.status, OrderStatus::Resting);
        assert_eq!(order.side, Side::Buy);
        assert!(order.error.is_none());

        assert!(!order.status.is_terminal());
        assert!(!order.status.is_rejected());
    }

    #[test]
    fn test_order_state_filled() {
        let json = r#"{
            "status": "filled",
            "symbol": "BTC-USD",
            "orderId": "xyz789",
            "price": 98000.0,
            "originalSize": 0.5,
            "size": 0.0,
            "filledSize": 0.5,
            "isBuy": true,
            "maker": false,
            "timestamp": 1770918312787284000
        }"#;

        let order: OrderState = serde_json::from_str(json).unwrap();

        assert_eq!(order.status, OrderStatus::Filled);
        assert!(order.status.is_terminal());
        assert!(!order.status.is_rejected());
        assert!(!order.is_maker);
        assert_eq!(order.filled_size, 0.5);
        assert_eq!(order.size, 0.0);
    }
}

File: crates/api-python/bulk_api/messages/account.py
import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional

from bulk_api.common import Side, OrderStatus


@dataclass
class OrderState:
    """Represents an order status update"""
    timestamp: int
    symbol: str
    order_id: str
    status: OrderStatus
    side: Side
    price: float
    vwap: float
    size: float
    size_done: float
    size_orig: float
    is_maker: bool
    error: Optional[str] = None

    def get_side(self) -> Side:
        """Get the order side"""
        return Side.BUY if self.is_buy else Side.SELL

    def amount_remaining(self) -> float:
        """Get the amount of remaining to be filled"""
        return max(self.size - self.size_done, 0.0)

    @classmethod
    def from_api(cls, data: Dict) -> 'OrderState':
        return cls(
            timestamp=data.get('timestamp'),
            symbol=data.get('symbol'),
            order_id=data.get('orderId'),
            status=OrderStatus.from_string(data.get('status')),
            side=Side.BUY if data.get('isBuy') else Side.SELL,
            price=data.get('price'),
            vwap=data.get('vwap'),
            size=data.get('size'),
            size_done=data.get('filledSize', 0.0),
            size_orig=data.get('originalSize', data.get('size')),
            is_maker=data.get('maker', False),
            error = data.get('reason')
        )

@dataclass
class Margin:
    """Account-level margin information"""
    total_balance: float
    available_balance: float
    margin_used: float
    notional: float
    realized_pnl: float
    unrealized_pnl: float
    fees: float
    funding: float

    @classmethod
    def from_api(cls, data: Dict) -> 'Margin':
        return cls(
            total_balance=data.get('totalBalance', 0.0),
            available_balance=data.get('availableBalance', 0.0),
            margin_used=data.get('marginUsed', 0.0),
            notional=data.get('notional', 0.0),
            realized_pnl=data.get('realizedPnl', 0.0),
            unrealized_pnl=data.get('unrealizedPnl', 0.0),
            fees=data.get('fees', 0.0),
            funding=data.get('funding', 0.0)
        )


@dataclass
class Position:
    """Position information with full risk metrics"""
    symbol: str
    size: float  # Positive for long, negative for short
    price: float  # Volume-weighted average price (entry price)
    fair_price: float  # Current fair/mark price
    notional: float  # size × fair_price
    realized_pnl: float
    unrealized_pnl: float
    leverage: float
    liquidation_price: float
    fees: float
    funding: float
    maintenance_margin: float
    lambda_: float  # Lambda value (risk parameter)
    risk_allocation: float  # C_i

    @classmethod
    def from_api(cls, data: Dict) -> 'Position':
        return cls(
            symbol=data.get('symbol', ''),
            size=data.get('size', 0.0),
            price=data.get('price', 0.0),
            fair_price=data.get('fairPrice', 0.0),
            notional=data.get('notional', 0.0),
            realized_pnl=data.get('realizedPnl', 0.0),
            unrealized_pnl=data.get('unrealizedPnl', 0.0),
            leverage=data.get('leverage', 0.0),
            liquidation_price=data.get('liquidationPrice', 0.0),
            fees=data.get('fees', 0.0),
            funding=data.get('funding', 0.0),
            maintenance_margin=data.get('maintenanceMargin', 0.0),
            lambda_=data.get('lambda', 0.0),
            risk_allocation=data.get('riskAllocation', 0.0)
        )

    def is_long(self) -> bool:
        """Check if position is long"""
        return self.size > 0

    def is_short(self) -> bool:
        """Check if position is short"""
        return self.size < 0


@dataclass
class LeverageSetting:
    """Leverage setting for a symbol"""
    symbol: str
    leverage: float  # 1.0 to 50.0

    @classmethod
    def from_api(cls, data: Dict) -> 'LeverageSetting':
        return cls(
            symbol=data.get('symbol', ''),
            leverage=data.get('leverage', 1.0)
        )


@dataclass
class AccountSnapshot:
    """Complete account state snapshot"""
    margin: Margin
    positions: List[Position]
    open_orders: List[OrderState]
    leverage_settings: List[LeverageSetting]
    timestamp: int = field(default_factory=lambda: int(time.time() * 1000))

    @classmethod
    def from_api(cls, data: Dict) -> 'AccountSnapshot':
        return cls(
            margin=Margin.from_api(data.get('margin', {})),
            positions=[Position.from_api(pos) for pos in data.get('positions', [])],
            open_orders=[OrderState.from_api(order) for order in data.get('openOrders', [])],
            leverage_settings=[LeverageSetting.from_api(lev) for lev in data.get('leverageSettings', [])]
        )

    def get_position(self, symbol: str) -> Optional[Position]:
        """Get position for a specific symbol"""
        for pos in self.positions:
            if pos.symbol == symbol:
                return pos
        return None

    def get_leverage_setting(self, symbol: str) -> Optional[LeverageSetting]:
        """Get leverage setting for a specific symbol"""
        for lev in self.leverage_settings:
            if lev.symbol == symbol:
                return lev
        return None


@dataclass
class MarginUpdate:
    """Real-time margin update"""
    total_balance: float
    available_balance: float
    margin_used: float
    notional: float
    realized_pnl: float
    unrealized_pnl: float
    fees: float
    funding: float

    @classmethod
    def from_api(cls, data: Dict) -> 'MarginUpdate':
        return cls(
            total_balance=data.get('totalBalance', 0.0),
            available_balance=data.get('availableBalance', 0.0),
            margin_used=data.get('marginUsed', 0.0),
            notional=data.get('notional', 0.0),
            realized_pnl=data.get('realizedPnl', 0.0),
            unrealized_pnl=data.get('unrealizedPnl', 0.0),
            fees=data.get('fees', 0.0),
            funding=data.get('funding', 0.0)
        )


@dataclass
class PositionUpdate:
    """Real-time position update"""
    symbol: str
    size: float
    price: float
    realized_pnl: float
    unrealized_pnl: float
    leverage: float
    liquidation_price: float
    fair_price: float
    notional: float
    fees: float
    funding: float
    maintenance_margin: float
    lambda_: float
    risk_allocation: float

    @classmethod
    def from_api(cls, data: Dict) -> 'PositionUpdate':
        return cls(
            symbol=data.get('symbol', ''),
            size=data.get('size', 0.0),
            price=data.get('price', 0.0),
            realized_pnl=data.get('realizedPnl', 0.0),
            unrealized_pnl=data.get('unrealizedPnl', 0.0),
            leverage=data.get('leverage', 0.0),
            liquidation_price=data.get('liquidationPrice', 0.0),
            fair_price=data.get('fairPrice', 0.0),
            notional=data.get('notional', 0.0),
            fees=data.get('fees', 0.0),
            funding=data.get('funding', 0.0),
            maintenance_margin=data.get('maintenanceMargin', 0.0),
            lambda_=data.get('lambda', 0.0),
            risk_allocation=data.get('riskAllocation', 0.0)
        )
